### ISSUE TITLE
Add a module to validate meta values during execution

### ISSUE THREAD

**Issue by deanebarker** [statiqdev/Statiq#210]
Some ideas:

```
// Make sure a specific key exists
ValidateMeta("Date") 

// Make sure it matches a format
ValidateMeta("Date").HasFormat("[0-9]{2}/[0-9]{2}/[0-9]{4}")

// Make sure it is the correct type
ValidateMeta("Date").IsType(typeof(DateTime))

// Make sure it can convert to the correct type (redundant from above?)
ValidateMeta("Date").CanConvertTo(typeof(DateTime))

// Make sure it fits in ranges
ValidateMeta("Age").CanConvertTo(typeof(Int32)).Between(0, 120)
ValidateMeta("Age").CanConvertTo(typeof(Int32)).GreaterThan(21)
ValidateMeta("Summary").ShorterThan(50)

// Maybe specify intelligent errors?
ValidateMeta("Summary").ShorterThan(50).WithError("Your summary is too short!")
```

If this check fails, I assume we should throw an very verbose exception listing the meta value and what was expected, then stop execution?  This would be the equivalent of a "pre-flight check," to make sure all the data is in order before we embark on the entire build.

I think this one requires a fair amount of thought, so just tossing around some ideas.

**Comment by daveaglick** [comment]
Might want to consider using an assertion library for this so that you don't have to come up with a full set of comparison logic. One suggestion would be to reference NUnit just for it's assertion classes - we're referencing NUnit for the unit tests in the solution anyway, so it's already being downloaded as a package. Alternativly, [FluentAssertions](http://www.fluentassertions.com/) is popular as is [Shouldly](https://github.com/shouldly/shouldly).

If you do use an assertions library, this module would need to go in a separate project to match our convention of no additional dependencies in the core library. I'm also fine if you think the range of checks wouldn't justify the extra dependency on an assertions library, in which case this could go in core.

**Comment by deanebarker** [comment]
I didn't even know an "assertion library" was a thing.  I'll check them out.

**Comment by deanebarker** [comment]
Ideally, here's how I'd like the syntax to look;

```
ValidateMeta("Age").Should().BeGreaterThan(4)
```

But I'm not sure how to do this.  `ValidateMeta` would invoke the module, but the syntax starting with the `Should()` would be a direct pass-thorugh to Fluent.

I could do this pretty easily...

```
ValidateMeta("Age", m => m.Should().BeGreaterThan(4)
```

...but that's noticeably less awesome.

Perhaps I'll start less awesome just to prove the theory, and become more awesome from there.

**Comment by deanebarker** [comment]
Also, I wonder how I might incorporate assertions to _all_ the meta at once.  For example, someone might want to assert that nothing _outside of a specified range of keys_ exists.  So, you can have keys A, B, and C, but not D.  In this case, we'd need to consider the actual `Metadata` collection as a whole, not just individual keys.

I wonder if we have different modules.  `ValidateMeta` and `ValidateAllMeta`, for example, which takes the `Metadata` collection.  Then perhaps even a `ValidateDocument` which takes the `IDocument` itself (or perhaps a `DocumentConfig`).

**Comment by daveaglick** [comment]
Don't fear - there's a standard way of solving this for fluent interfaces :smile: I'll write up some details once I'm in the office.

**Comment by deanebarker** [comment]
Second thoughts...

Perhaps using FluentAssertions is the "advanced" version of this? Perhaps we build a smaller validation library into the core.  Honestly, 99% of validation would be:
1. Ensure a key exists
2. Ensure it can convert to the proper type

So, it would look like this:

```
ValidateMeta("Age")
ValidateMeta("Age", typeof(Int32)
```

Additionally, we could create a `Func<object, bool>` that took the value and returned a boolean.

```
ValidateMeta("Age", o => o > 21)
```

Perhaps FluentAssertions is wrapped for a separate module?  I could build the above into the core in about 30 minutes, with no dependencies.  And we could always save the FluentAssertions for another module for people needing more advanced functionality.

```
FluentAssertion("Age").This().Is().Some().Amazingly().Fluent().Assertion()
```

The trick, of course, is managing the line between them. You don't want creeping functionality in the core module.  You need some "jumping off" point where you say, "Get the FluentAssertions module for that..."

**Comment by deanebarker** [comment]
Two fluent methods:

```
.DelimitedBy(",")
```

This would split the value and perform the validation on each item.  Handy for comma-delimited numeric values: `Chapters: 1,2,3`

```
.WithMessage("{0} must be greater than 21")
```

This would output on failure, with the meta key replacing the `{0}`.

**Comment by daveaglick** [comment]
Let's address the fluent interface pattern first. The trick here is to return a special class from the first (and possibly subsequent) methods in the chain instead of the actual thing you want to operate on. I've sometimes seen this class referred to as a "data context class", but really it's just a wrapper around your main thing plus some extra information.

For example:

```
public class ValidateMeta : IModule
{
    private readonly string _key;
    private readonly Func<object, bool> _condition;

    public ValidateMeta(string key)
    {
        _key = key;
        _condition = _ => true;  // Return true by default until we use the fluent interface
    }

    // This could be used at the end of the fluent chain or directly for more control
    public ValidateMeta(string key, Func<object, bool> condition)
    {
        _key = key;
        _condition = condition;
    }

    public IValidateMetaShouldContext Should()
    {
        return new ValidateMetaShouldContext(_key, true);
    }

    public IValidateMetaShouldContext ShouldNot()
    {
        return new ValidateMetaShouldContext(_key, false);
    }

    // Hide the implementation so users can't create it directly
    private class ValidateMetaShouldContext
    {
        private readonly string _key;
        private readonly bool _should;

        public ValidateMetaContext(string key, bool should)
        {
            _key = key;
            _should = should;
        }

        // Return IModule if this is the last method in the chain
        IModule BeGreaterThan(int n)
        {
            // Condition would probably need to be more sophisticated, I.e., which
            // type are we trying to compare (maybe use a generic context class?)
            return new ValidateMeta(_key, x => ((int)x) > n);
        }
    }

    // Execute method would get the metadata value and evaluate _condition,
    // throwing if the condition doesn't match    
}

public interface IValidateMetaShouldContext
{    
    IModule BeGreaterThan(int n);
}
```

The different context classes let you enforce a specific series of chains by only allowing certain fluent methods at certain stages of the method chain. For example, we shouldn't be able to write `ValidateMeta("Age").BeGreaterThan(4).Should()` which would be allowed if a single class contained all the fluent methods.

There's one problem with this though - you'd normally run into [the finishing problem](http://daveaglick.com/posts/method-chaining-fluent-interfaces-and-the-finishing-problem). That is, you're returning a context class but the pipeline is expecting an `IModule` at the end of the method chain. That's why I made the methods in `IValidateMetaShouldContext` instantiate a new `ValidateMeta` class and return an `IModule`. The `Func<object, bool>` actually encapsulates all of the information our context class(es) had been carrying for us in the form of a function. A nice side benefit is that by exposing the alternate constructor that the fluent interface uses to pass in a function, users can also pass in their own more complicated functions using whatever assertions library they want to.

**Comment by daveaglick** [comment]
> Perhaps using FluentAssertions is the "advanced" version of this

After specing out the example above, I'm inclined to agree. In fact, I don't think we need to provide the "advanced" version. As long as we have a constructor in `ValidateMeta` that accepts a delegate, I would say let the user make use of whatever library they want. At that point it's up to them to take the value we've fetched for their key (assuming one existed in the first place) and return a bool.

This does bring up an additional point though, since many assertion libraries actually throw on failure (instead of returning a bool), we'll probably want to wrap the call to the `_condition` function in `Execute()` with a try-catch so that we can substitute whatever additional information we need to in the eventual error message.

**Comment by daveaglick** [comment]
Also, love the idea of providing an optional `.WithMessage()` method for customizing the message.

**Comment by deanebarker** [comment]
Here's an important point I think we need to acknowledge --

The intention of this is really only to validate meta stemming from front matter.  Meta coming in from YAML in a Markdown file is "loose."  It's not validated when the document is saved, and can blow up a pipeline.  _This_ is what needs to be validated.

Meta that gets added programmatically in pipelines isn't nearly as critical.  The programmer added this, so there's a much higher assumption that it's just fine.  It could be wrong, sure, but if it is, this is something that needs to be corrected in the config file, really.

The two major things we need to ensure about "front matter meta":
1. It exists
2. It's in the right format

Maybe this is obvious, but this is what led me to "walk back" from the Fluent Assertions plan.  I still think that would be a great advanced library, but it's overkill for the immediate problem which I describe here.

**Comment by deanebarker** [comment]
Your point about the delegate is true.  This code I had above:

```
ValidateMeta("Age", m => m.Should().BeGreaterThan(4)
```

This would work fine with or without a formal Fluent Assertions library (assuming we trapped the Exception, of course).

**Comment by daveaglick** [comment]
> it's overkill for the immediate problem which I describe here

Agree completely, I would just stick to a simple set of assertions for now.

> only to validate meta stemming from front matter

I do really like the module being totally source agnostic though. Even if the current use case is primarily for metadata from front matter, I'd guess there are probably lots of use cases out there where validating metadata from other sources would be valuable (I.e., I pull it from a web resource and want to make sure it's good before I publish).

Another thing - in thinking about typing, I wonder if making this module generic would help. The config file processor is already set up to handle generic modules, we just haven't had an occasion to use them much. Then you would write:

```
ValidateMeta<int>("Age").Should().BeGreaterThan(4);
```

...or...

```
ValidateMeta<int>("Age", age => age > 4);
```

But the beauty of making it generic is that the function input will already be typed. More advanced example:

```
ValidateMeta<IEnumerable<int>>("Tires", tires => tires.Count() >= 4);
```

**Comment by deanebarker** [comment]
Do you have an example of a generic module?  This is where my "not a developer" schtick is very true.  I understand generics, but not enough to write a module as one.

**Comment by daveaglick** [comment]
No problem - here's a more complete example with generics:

```
public class ValidateMeta<T> : IModule
{
    private readonly string _key;
    private readonly Func<T, bool> _condition;

    public ValidateMeta(string key)
    {
        _key = key;
        _condition = _ => true;  // Return true by default until we use the fluent interface
    }

    // This could be used at the end of the fluent chain or directly for more control
    public ValidateMeta(string key, Func<T, bool> condition)
    {
        _key = key;
        _condition = condition;
    }

    public IValidateMetaShouldContext<T> Should()
    {
        return new ValidateMetaShouldContext(_key, true);
    }

    public IValidateMetaShouldContext<T> ShouldNot()
    {
        return new ValidateMetaShouldContext(_key, false);
    }

    // Hide the implementation so users can't create it directly
    private class ValidateMetaShouldContext : IValidateMetaShouldContext<T>
    {
        private readonly string _key;
        private readonly bool _should;

        public ValidateMetaShouldContext(string key, bool should)
        {
            _key = key;
            _should = should;
        }

        // Return IModule if this is the last method in the chain
        public IModule BeGreaterThan(T n, Comparer<T> comparer = null)
        {
            comparer = comparer ?? Comparer<T>.Default;
            return new ValidateMeta<T>(_key, metadata => comparer.Compare(metadata, n) > 0);
        }
    }

    public IEnumerable<IDocument> Execute(IReadOnlyList<IDocument> inputs, IExecutionContext context)
    {
        return inputs.AsParallel().Select(input =>
        {
            // Get the metadata value converted to T (the built-in metadata conversion stuff will handle this)
            T value;
            if (!input.MetadataAs<T>().TryGetValue(_key, out value))
            {
                // Key didn't exist or couldn't be converted to T
            }

            // Check the condition, wrapping in a try-catch in case the check throws
            try
            {
                if (!_condition(value))
                {
                    // Condition failed
                }
            }
            catch (Exception ex)
            {
                // Condition threw an exception
            }

            return input;
        });
    }
}

public interface IValidateMetaShouldContext<T>
{
    IModule BeGreaterThan(T n, Comparer<T> comparer = null);
}
```

Notice how using generics neatly gets us around the problem of needing to worry about the type of the metadata and the type of the comparison value - we just let .NET use the default comparison logic for the specified generic type and allow the user to substitute an alternative if they want.

**Comment by deanebarker** [comment]
But here's the problem: all meta coming in from Yaml in Markdown is going to be a string.  With those, we assume that it's a string, and what we want to do _is make sure we can do is convert it to something else when we have to_.  That's the big thing -- this meta might not be T, but can I make it T when I need to?

**Comment by daveaglick** [comment]
Yep - that's probably one of the totally unique things about Wyam. The internal metadata structure stores the original value, but when I'm getting metadata out I don't really care about the type of the original value. If I stored a string (as in YAML) and I want a number, that'll get handled for me. The conversion operations are very lenient. That's why, given the exact same document with a metadata key "Age" that's storing a `string` of "4", I can write all of the following in a Razor template and they won't fail:

```
@{
    var originalAge = Model.Get("Age");  // originalAge.GetType() == string, value == "4"
    var ageAsInt = Model.Get<int>("Age");  // ageAsInt.GetType() == int, value == 4
    var ageAsDouble = Model.Get<double>("Age");  // ageAsDouble.GetType() == double, value = 4.0
    var lotsOfIntAges = Model.Get<IEnumerable<int>>("Age");  // lotsOfIntAges.GetType() == IEnumerable<int>, value = []{ 4 }
    // Etc.
}
```

Since the sample above was using the following line to get the metadata value as whatever type `T`:

```
input.MetadataAs<T>().TryGetValue(_key, out value)
```

...the metadata conversion process will attempt to convert it to the requested type, and probably succeed if the conversion is reasonable. This flexibility would be important because if I'm getting a string "4" from YAML front matter but I want to assert "age should be greater than 3", that question could have a totally different answer when comparing the strings `"3"` and `"4"` than comparing the ints `3` and `4`. And it may actually succeed if the value is `"Unknown"` when you really meant to do a numeric comparison which should fail because it's not a number.

**Comment by daveaglick** [comment]
> what we want to do is make sure we can do is convert it to something else when we have to

The generic module would take care of this because it's using the same metadata conversion logic as everything else. If that `input.MetadataAs<T>().TryGetValue(key, out value)` call returns false, then you know the metadata logic doesn't know how to convert whatever is in the metadata key to the requested type `T`. In other words, in addition to performing the more specific check you define with the fluent interface or delegate, a generic ValidateMeta<T> module would also perform an implicit validation that the metadata can be converted to T when requested.

**Comment by deanebarker** [comment]
Okay, I'm working through this.  I have a fairly solid implementation going.

However, is there a cleaner way to shut down execution, other than spitting back in unhandled exception?  It just seems messy, and honestly, it's hard to find the actual error message inside the stacktrace.

**Comment by daveaglick** [comment]
:thumbsup:

> is there a cleaner way to shut down execution, other than spitting back in unhandled exception

Not right now - I updated it recently to make sure all exceptions killed processing, but we could probably also add a `IExecutionContext.Abort(string message)` method. Might be a few days until I can add it though - almost done with the document factory stuff and really want to get that wrapped up.

**Comment by deanebarker** [comment]
Okay, I'll throw 'em ugly for the time being...

**Comment by deanebarker** [comment]
I'm thinking about configuring the check conditions as fluent methods, since it might be helpful to have more than one, and each would have a unique error message.  By comparison, neither the (1) existence check, or the (2) type check, need custom error messages because they're pretty straightforward: the thing wasn't found, or it couldn't be converted.

I'm thinking it might look something like this:

```
ValidateMeta<string>("Title")
   .WithCondition(s => s.Length > 5, "Title must be greater than 5 characters")
   .WithCondition(s => s.Contains("Gumby"), "Title must contain the word \"Gumby\"")
   .WithCondition(s => s.EnsureNotAllCaps(), "Stop yelling, you idiot")
```

There are _five_ checks there: (1) the check for existence, (2) the check for type, (3-5) the checks for content.  The content checks will each have unique error messages.

Also: `WithCondition` or `AddCondition`.  I don't care. What sounds more correct?

**Comment by daveaglick** [comment]
That's really coming along - I like this design a lot. 

I'm fond of the "With...()" syntax for fluent methods that add additional state, and that seems to follow what little convention there is around fluent APIs.

**Comment by deanebarker** [comment]
Explain to me the difference between:

```
ValidateMeta<int>("Age")
ValidateMeta("Age", typeof<int>)
```

I started from your code, and I'm pretty sure I understand it.  It seems the `<int>` is just a different way to pass in a type parameter. Don't get me wrong -- I understand that generics have a valid purpose, but _in this particular instance_ it seems that we're just using it to pass in a type.

I don't have a problem with this, but (1) it forces a type check, when the user may not want or need one, and (2) it makes this module different from every other module.

While generics are slick, is it providing enough value here to compensate for the difference in format with other modules?

**Comment by daveaglick** [comment]
The difference is that when you pass in a `Type` object, that's just another object instance like any other that happens to describe type information. When you use a generic type parameter, `T`, that's actually baked into the class by the compiler. Behind the scenes, the compiler creates seperate definitions for your class depending on the generic type parameters you used. This has lots of advantages, but for our purpose the main one is that you can have strongly-typed delegates now.

Consider the `WithCondition()` method. Without the generic type parameter, you'll have to use a `Func<object, bool>` for the condition. That means that when you're specifying conditions in the config file, you'll have to cast to the proper type inside every condition (unless your check happens to operate on `object`):

```
.WithCondition(s => ((string)s).Length), "Message...")
```

Perhaps more importantly, when you get the value from the metadata collection in the `Execute()` method, there's no way to tell the metadata collection to give you an object converted to a specific type represented by a `Type` object. That whole mechanism is based on generics, and changing it to support reflection-based conversions is non-trivial. So you'll end up only getting the original type of the metadata value and won't be able to test if the metadata collection can properly convert to the needed type.

I wouldn't worry too much about this being "different". I always envisioned generic modules and there are unit tests to make sure they're supported correctly in the config.

**Comment by deanebarker** [comment]
I see, I see.  I missed this line of code:

```
Func<T, bool> _condition;
```

That makes sense now.  Thanks.  I'll leave it as a generic.

[sigh]  Someday I'll be a _real_ developer...

![](https://media0.giphy.com/media/gOCvwLSvGLmus/200_s.gif)

**Comment by daveaglick** [comment]
Generics are not the easiest concept to grasp, and there are lots of professional developers who have no idea how they work. That your using them here at all makes you a "real" developer by any measure!

**Comment by deanebarker** [comment]
Should the condition check be positive or negative?  Consider:

```
   .WithCondition(s => s.Length > 5, ...
```

Should we consider the condition violated if that returns `true` or `false`?  Put another way, is the condition a test for _validity_ (true if it's greater than 5 characters) or a test for _invalidity_ (false if it's greater than 5 characters).

Technically, it's just a boolean switch.  But what makes more sense logically and semantically?

**Comment by daveaglick** [comment]
Good point - maybe the method should be `WithAssertion()`? That would make it clear that if the delegate returns false, there's a problem.

**Comment by deanebarker** [comment]
I have it running. I would like to write some unit tests for it, but I reviewed some of your testing code, and I can't make heads or tails of it.  Is there a simple example somewhere?

**Comment by daveaglick** [comment]
Haha. Hell, I usually can't make heads or tails out of my testing code. This set of tests probably isn't a bad starting point:

https://github.com/Wyamio/Wyam/blob/develop/src/Wyam.Modules.Json.Tests/GenerateJsonTests.cs

In general, the test is going to look like this:

```
[TestFixture]
[Parallelizable(ParallelScope.Self | ParallelScope.Children)]
public class ModuleNameTests : BaseFixture
{
    public class ExecuteMethodTests : ModuleNameTests
    {
        [Test]
        public void ShortDescriptionOfWhatImTesting()
        {
            // Given
            // Set up a mock for IDocument and IExecutionContext here
            // Also create an instance of your module

            // When
            myModule.Execute(new[] { document }, context).ToList();  // Make sure to materialize the result list

            // Then
            // Check that results are as expected and that your mocks got the calls they should have
        }
    }
}
```

I'm using [NSubstitute](http://nsubstitute.github.io/) for the mocking which makes it pretty easy to create the mocks, register behavior for certain method calls, and check for received calls.

Just let me know if you need any more help along the way. You could also submit a PR and I'll make an initial test or two that you could use as a template for additional tests, and then just submit a second PR.

**Comment by deanebarker** [comment]
I successfully wrote a test, but Visual Studio isn't recognizing them. Are you running them out of Visual Studio, or do you run NUnit tests some other way?

**Comment by daveaglick** [comment]
Personally, I use the Resharper test runner. Wyam is on the new NUnit 3 release, so you might need the updated test adapter to get the Visual Studio runner to see them:

https://github.com/nunit/nunit3-vs-adapter/wiki

https://visualstudiogallery.msdn.microsoft.com/0da0f6bd-9bb6-4ae3-87a8-537788622f2d

**Comment by deanebarker** [comment]
The pull request is in.  I wrote one test, which I also submitted, but I'm having trouble getting Visual Studio to run it, and I don't quite understand NSubstitute, so that's on me to figure out.

Also, I learned from my last mistake and created a separate branch for this.

![](https://misscraftyengineer.files.wordpress.com/2013/08/billy-madison-i-am-the-smartest-man-alive.gif?w=470&h=238)

**Comment by daveaglick** [comment]
:sparkles: 

Closed by #220
